src/Checkout/Payment/Method/AbstractPaymentMethodHandler.php

```diff
@@ -7,10 +7,302 @@
+        Context $context,
+        ?Struct $validateStruct,
+    ): ?RedirectResponse {
+        $paypalOrderId = $request->request->getAlnum(self::PAYPAL_PAYMENT_ORDER_ID_INPUT_NAME);
+        [$orderTransaction, $order] = $this->fetchOrderTransaction($transaction->getOrderTransactionId(), $context);
+        $existingVault = $this->isVaultable() ? $this->vaultTokenService->getAvailableToken($transaction, $orderTransaction, $order, $context) : null;
+        if ($this->requirePreparedOrder() && !$paypalOrderId && !$existingVault) {
+            throw CheckoutException::preparedOrderRequired(static::class);
+        }
+
+        $this->settingsValidationService->validate($order->getSalesChannelId());
+
+        $this->stateMachineRegistry->transition(
+            new Transition(
+                OrderTransactionDefinition::ENTITY_NAME,
+                $transaction->getOrderTransactionId(),
+                $this->getProgressTransactionState(),
+                'stateId'
+            ),
+            $context
+        );
+
+        $response = null;
+        if (!$paypalOrderId) {
+            $paypalOrder = $this->orderBuilder->getOrder($transaction, $orderTransaction, $order, $context, $request);
+            $response = $this->orderResource->create(
+                $paypalOrder,
+                $order->getSalesChannelId(),
+                $this->resolvePartnerAttributionId($request),
+                true,
+                $transaction->getOrderTransactionId() . ($orderTransaction->getUpdatedAt()?->getTimestamp() ?: ''),
+                $this->getMetaDataId($request),
+            );
+            $paypalOrderId = $response->getId();
+        }
+
+        $this->transactionDataService->setOrderId(
+            $transaction->getOrderTransactionId(),
+            $paypalOrderId,
+            $this->resolvePartnerAttributionId($request),
+            $order->getSalesChannelId(),
+            $context,
+        );
+
+        if (!$response) {
+            $this->orderPatchService->patchOrder(
+                $order,
+                $orderTransaction,
+                $context,
+                $paypalOrderId,
+                PartnerAttributionId::PAYPAL_PPCP
+            );
+            $response = $this->orderResource->get($paypalOrderId, $order->getSalesChannelId());
+        }
+
+        if ($action = $this->resolveRedirect($response)) {
+            return new RedirectResponse($action);
+        }
+
+        $this->executeOrder(
+            $transaction,
+            $response,
+            $order,
+            $orderTransaction,
+            $context,
+        );
+
+        return null;
+    }
+
+    protected function executeOrder(PaymentTransactionStruct $transaction, Order $paypalOrder, OrderEntity $order, OrderTransactionEntity $orderTransaction, Context $context, bool $isUserPresent = true): Order
+    {
+        if (!$this->attemptExecution()) {
+            return $paypalOrder;
+        }
```

Comment:
can't we override this function and add the early return instead?

Reply:
yes, APMHandler and PUIHandler will now override `executeOrder` instead.